dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java

```diff
@@ -67,32 +70,62 @@ public List<BlockWorkerInfo> getPreferredWorkers(List<BlockWorkerInfo> blockWork
   private static class ConsistentHashProvider {
     private static final HashFunction HASH_FUNCTION = murmur3_32_fixed();
     private static final int MAX_ATTEMPTS = 100;
-    private List<BlockWorkerInfo> mLastWorkerInfos = ImmutableList.of();
-    private NavigableMap<Integer, BlockWorkerInfo> mActiveNodesByConsistentHashing;
-
-    private volatile long mLastUpdatedTimestamp = 0L;
-
-    private final AtomicBoolean mNeedUpdate = new AtomicBoolean(false);
-
-    private static final long WORKER_INFO_UPDATE_INTERVAL_MS = 1000L;
-
+    private static final long WORKER_INFO_UPDATE_INTERVAL_NS = Constants.SECOND_NANO;
+    private volatile List<BlockWorkerInfo> mLastWorkerInfos = ImmutableList.of();
+    // Can only be null before the first call to refresh
+    @Nullable
+    private volatile NavigableMap<Integer, BlockWorkerInfo> mActiveNodesByConsistentHashing;
+    private final Semaphore mInitLock = new Semaphore(1);
+
+    // Must use System.nanoTime to ensure monotonic increment
+    private final AtomicLong mLastUpdatedTimestamp = new AtomicLong(System.nanoTime());
+
+    /**
+     * Initializes or refreshes the worker list using the given list of workers and number of
+     * virtual nodes.
+     */
     public void refresh(List<BlockWorkerInfo> workerInfos, int numVirtualNodes) {
-      // check if we need to update worker info
-      if (mLastUpdatedTimestamp <= 0L
-          || System.currentTimeMillis() - mLastUpdatedTimestamp > WORKER_INFO_UPDATE_INTERVAL_MS) {
-        mNeedUpdate.set(true);
+      // When the active nodes map does not exist, the hash provider is not initialized yet.
+      // let one caller initialize the map while blocking all others.
```

Comment:
this begs the question: what if multiple callers want to initialize with different set of workers?
I think either we make `ConsistentHashProvider` non static and one instance per `FileSystemContext`, or the responsibility of initializing and refreshing it should go to the membership service, which monitors and maintains a consistent view of available workers.

Reply:
I think a singleton per `FileSystemContext` sounds good